### ISSUE TITLE
Optimize trip shape buffer creation

### ISSUE THREAD

**Issue by barbeau** [CUTR-at-USF/gtfs-realtime-validator#199]
**Summary:** 

https://github.com/CUTR-at-USF/gtfs-realtime-validator/pull/194 added new rule `E029 - Vehicle position outside trip shape buffer`, based on shapes.txt if provided.

The downside is that building the buffered trip shapes in `GtfsMetadata` constructor (executed on each GTFS dataset the first time during execution that GTFS dataset is encountered - in other words, it has to load the GTFS Metadata into memory) takes a substantial amount of time - it increased the time required to build the HART metadata from around 1 second to 15 seconds - sample log output following implementation:

~~~
INFO edu.usf.cutr.gtfsrtvalidator.api.resource.GtfsRtFeed - http://api.tampa.onebusaway.org:8088/vehicle-positions is a valid GTFS-realtime feed
GtfsMetadata - Building GtfsMetadata for http://gohart.org/google/google_transit.zip...
GtfsMetadata - Loaded shapes.txt points for http://gohart.org/google/google_transit.zip
GtfsMetadata - Generated shapes.txt bounding boxes for http://gohart.org/google/google_transit.zip
GtfsMetadata - Sorting shape points for http://gohart.org/google/google_transit.zip...
GtfsMetadata - Shape points for http://gohart.org/google/google_transit.zip are sorted.
GtfsMetadata - Processing trips and building trip shapes for http://gohart.org/google/google_transit.zip...
GtfsMetadata - Trips and trip shape buffers processed for http://gohart.org/google/google_transit.zip in 14.14205 seconds
GtfsMetadata - Built GtfsMetadata for http://gohart.org/google/google_transit.zip in 14.14439 seconds
~~~

We should look to see if there is a way to optimize building these trip shape buffers, or at least offer a way to turn off this rule/metadata via command-line option.

The good news is that we only pay the penalty once for each runtime per feed so it's not the end of the world, although this can be a bit painful during development when trying to implement and test new rules.

I'm also not exactly sure what impact adding this trip shape buffer metadata had on memory - that's something to look at as well.

**Steps to reproduce:** 

Run the validator and load HART's feed:

* GTFS-rt - http://api.tampa.onebusaway.org:8088/vehicle-positions
* GTFS - http://gohart.org/google/google_transit.zip

**Expected behavior:** 

Metadata should build relatively quickly, or give me an option not to wait for metadata related to a single rule

**Observed behavior:** 

You have to wait around 15 seconds for the initial metadata build for a reasonably-sized feed

**Platform:** 

Windows 7 Enterprise w/ jdk1.8.0_73 and Chrome Version 58.0.3029.110 (64-bit)

**Comment by barbeau** [comment]
So performance on MBTA data is significantly worse - building metadata takes 104 seconds:

~~~
GtfsMetadata - Building GtfsMetadata for http://www.mbta.com/uploadedfiles/MBTA_GTFS.zip...
GtfsMetadata - Processing trips and building trip shapes for http://www.mbta.com/uploadedfiles/MBTA_GTFS.zip...
GtfsMetadata - Trips and trip shape buffers processed for http://www.mbta.com/uploadedfiles/MBTA_GTFS.zip in 103.103577 seconds
GtfsMetadata - Built GtfsMetadata for http://www.mbta.com/uploadedfiles/MBTA_GTFS.zip in 104.104076 seconds
~~~

We need to figure out some way to reduce this.

**Comment by barbeau** [comment]
I took an initial shot at optimization in WIP PR https://github.com/CUTR-at-USF/gtfs-realtime-validator/pull/200 by eliminating trip shape building for trips with a service date from calendar.txt that does not include the current date.  This cut the metadata build time from around 104 seconds to 38 seconds, but has some problems - see https://github.com/CUTR-at-USF/gtfs-realtime-validator/pull/200 for details.

**Comment by barbeau** [comment]
It looks like creating the buffer around the line is the operation that takes the most amount of time - if I create the LineString without the buffer, the MBTA feed builds in under 5 seconds (for all trips, not just active ones):

~~
GtfsMetadata - Building GtfsMetadata for http://www.mbta.com/uploadedfiles/MBTA_GTFS.zip...
GtfsMetadata - Processing trips and building trip shapes for http://www.mbta.com/uploadedfiles/MBTA_GTFS.zip...
GtfsMetadata - Trips and trip shape buffers processed for http://www.mbta.com/uploadedfiles/MBTA_GTFS.zip in 4.4059 seconds
GtfsMetadata - Built GtfsMetadata for http://www.mbta.com/uploadedfiles/MBTA_GTFS.zip in 4.4445 seconds
~~

**Comment by barbeau** [comment]
After working with this a bit, the best solutions seems to be to build the buffers dynamically as the vehicle positions and trip assignments are discovered in the GTFS-rt feed.  I've implemented this in https://github.com/CUTR-at-USF/gtfs-realtime-validator/commit/3f44cf95ea9a213e190bd9bb55308fbf8f05e3e7 - from that commit summary:

>* Instead of building all buffered trip shapes when building the initial feed GtfsMetadata (as implemented in #194), now we build the buffered trip shapes on-demand based on the vehicles appearing in the GTFS-rt feed in real-time and their current trip assignment.  This has the advantage of always remaining relevant (versus trying to filter out trips that aren't relevant based on the GTFS calendar.txt and calendar_dates.txt for the current date - see #200 for an initial stab at that), and only builds the shapes that are needed, saving memory as well.
